### PR TITLE
Skip gradient updates when grad norm exceeds threshold

### PR DESCRIPTION
**Pull request by yfw**
# What does this PR do ?
Add `grad_norm_skip_threshold` config (default `float('inf')`) to skip gradient updates when the gradient norm is too large. The optimizer does not take a step if `grad_norm > grad_norm_skip_threshold` (as opposed to taking a zero-ing out the gradients)

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/optimizer/optimizer_config.py [hunk 1]
```diff
@@ -371,6 +371,9 @@ class OptimizerConfig:
     clip_grad: float = 1.0
     """Gradient clipping based on global L2 norm."""
 
+    grad_norm_skip_threshold: float = 1000
+    """Skip gradient update if the gradient norm exceeds this threshold."""
```
Comment (wujingyue) [inline megatron/core/optimizer/optimizer_config.py]: Is it a common recipe to deal with grad norm overflow? Any reference you can add to the PR for posterity?

**Review by wujingyue — COMMENTED** [review]
LGTM otherwise

File: megatron/core/optimizer/optimizer_config.py (hunk 1 above)
Comment (macandro96) [reply megatron/core/optimizer/optimizer_config.py]: @yfw is there a reference that we can add here? I believe it was requested from the research team for experimentation?

Comment (yfw) [reply megatron/core/optimizer/optimizer_config.py]: Yes, this was added to stabilize super and ultra training. CC @gshennvm who had added this originally and may be able to provide more context.

Comment (macandro96) [reply megatron/core/optimizer/optimizer_config.py]: @gshennvm please feel free to link details to any experiments if applicable; @wujingyue let me know if you need any more context?

**Review by wujingyue — APPROVED** [review]

File: megatron/core/optimizer/optimizer_config.py (hunk 1 above)
Comment (wujingyue) [reply megatron/core/optimizer/optimizer_config.py]: LGTM but please try your best to add context in code comment for posterity. Knobs like this can quickly drift out of place unless the codebase explains why they’re needed. Thanks!

Comment (macandro96) [reply megatron/core/optimizer/optimizer_config.py]: Will do!

File: megatron/core/optimizer/optimizer.py [hunk 2]
```diff
@@ -1463,9 +1468,15 @@ def step(self):
...
+                log_single_rank(
+                    logger, logging.INFO, "skipping grad norm because it's too large %s", grad_norm
+                )
+                use_decoupled_grad = optimizer.config.use_precision_aware_optimizer_no_fp8_or_ds_fp8
+                zero_grads_manual_impl(parameters, use_decoupled_grad=use_decoupled_grad)
+
         # Count the zeros in the grads.
         num_zeros_in_grad = self.count_zeros() if self.config.log_num_zeros_in_grad else None
-
         update_successful = self.step_with_ready_grads()
```
Comment (deepakn94) [inline megatron/core/optimizer/optimizer.py]: Update `update_successful` to `False` if the grad_norm is sufficiently large?

Comment (macandro96) [reply megatron/core/optimizer/optimizer.py]: Makes sense. Added tests for this too.

**Review by yaoyu-33 — APPROVED** [review]

File: megatron/core/optimizer/optimizer.py [hunk 3]
```diff
@@ -1463,10 +1469,17 @@ def step(self):
                     ),
                 )
 
+            if grad_norm > optimizer.config.grad_norm_skip_threshold:
+                log_single_rank(
+                    logger, logging.INFO, "skipping grad norm because it's too large %s", grad_norm
+                )
+                use_decoupled_grad = optimizer.config.use_precision_aware_optimizer_no_fp8_or_ds_fp8
+                zero_grads_manual_impl(parameters, use_decoupled_grad=use_decoupled_grad)
+                should_skip_update = True
```
Comment (deepakn94) [inline megatron/core/optimizer/optimizer.py:1471]: Actually if we skip the update below, then do we need to zero grads?

Comment (macandro96) [reply megatron/core/optimizer/optimizer.py:1471]: Good point; we don't need to. I removed the `zero_grads_manual_impl` function as it will not be used anywhere. Also - changed the default value of `grad_norm_skip_threshold` to `float('inf')` as this can be set in the config if the behavior is desired.

cc: @yfw

**Review by deepakn94 — APPROVED** [review]
